### PR TITLE
fix: implement missing getSize method in SharedFile.php

### LINKED ISSUE #36741: Webdav public files API issue with file and folder sizes
### Steps to reproduce

1. Setup [Phoenix](https://github.com/owncloud/phoenix) which uses said API
1. Inside of Phoenix
1. Create a folder "test"
1. Open the link share panel on "test"
1. Create a public link with "Editor" role (which will allow viewing and uploading)
1. Open public link in Phoenix
1. Create a folder "sub"
1. Enter the "sub" folder
1. Upload a few files with non-zero size
1. Refresh the page and check the size column
1. Go back to the parent folder and check the size of the "sub" folder

### Expected result
Correct size for the uploaded files.
Correct combined size for the "sub" folder.
In the Webdav response:
- folder size must be in "oc:size"
- file size must be in "dav:getcontentlength"

### Actual result
Folder size is correct but all files appear with 0 bytes size.
In the Webdav response:
- folder **and** file size is returned in "oc:size"
- "dav:getcontentlength" is always zero

See:
image

I had a quick look at the Sabre plugin but didn't find any obvious issue there. The problem is likely hidden in some implicit bits of Sabre DAV.

@DeepDiver1975

### PR DESCRIPTION
**Pull request by C0rby**
## Description

Implement the method `getSize` in `SharedFile.php`

## Related Issue

- Fixes #36741

## Motivation and Context

The webdav-api for public files couldn't get the correct file size
since `SharedFile.php` did not implement the `getSize` method.

## How Has This Been Tested?

To test execute a PROPFIND request for a shared file via the webdav public-files API
`curl -s --user user:password --basic http://localhost:8080/remote.php/dav/public-files/HXEGrChNif8AWBq -X PROPFIND`

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Database schema changes (next release will require increase of minor version instead of patch)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [x] Code changes
- [x] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link> 
- [x] Changelog item, see [TEMPLATE](https://github.com/owncloud/core/blob/master/changelog/TEMPLATE)

### CODE REVIEW

**Comment by mmattel** [comment]
@phil-davis this might be of interest regarding tests.
Is this something that should be included in 10.4 or must wait for a next release?

**Comment by mmattel** [comment]
@phil-davis @micbar added you to the reviewers.

**Comment by phil-davis** [comment]
@micbar do we try and do this for 10.4?

If so, then could add an acceptance test to cover the expected `PROPFIND` response for such public link requests.

**Comment by phil-davis** [comment]
squashed the commits and rebased

**Comment by phil-davis** [comment]
Unassigned myself and added the QA-team label.
@dpakach is going to allocate someone to do acceptance test(s)

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 1]
```diff
@@ -1099,6 +1099,58 @@ public function downloadFileAsUserUsingPassword(
...
+	 * @param $resource
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function userGetsSizeOfFileSharedByPublicLinkUsingTheWebdavApi($user, $resource) {
+		$tokenArray = $this->getLastShareData()->data->token;
+		$token = (string)$tokenArray[0];
+		$headers = $this->guzzleClientHeaders;
+		$password = $this->getPasswordForUser($user);
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php]: This should not need a user and password.
There will other code to "copy from" that accesses public links without needing a user/password.

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 2]
```diff
@@ -1099,6 +1099,58 @@ public function downloadFileAsUserUsingPassword(
...
+		$url = $this->getBaseUrl() . "/remote.php/dav/public-files/{$token}";
+		$this->response = HttpRequestHelper::sendRequest(
+			$url, "PROPFIND", $user, $password, $headers, null, null, null
+		);
+	}
+
+	/**
+	 * @When user :user gets size of file :resource using the WebDAV API
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php]:
```suggestion
	 * @When user :user gets the size of file :resource using the WebDAV API
```

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 3]
```diff
@@ -1099,6 +1099,58 @@ public function downloadFileAsUserUsingPassword(
...
+	 *
+	 * @param $size
+	 *
+	 * @return void
+	 */
+	public function theSizeOfTheFileShouldBe($size) {
+		$responseXml = HttpRequestHelper::getResponseXml($this->response);
+		$responseXml->registerXPathNamespace('d', 'DAV:');
+		$xmlPart = $responseXml->xpath("//d:prop/d:getcontentlength");
+		Assert::assertEquals($size, (string)$xmlPart[0]);
```
Comment (phil-davis) [inline tests/acceptance/features/bootstrap/WebDav.php]:
```suggestion
		Assert::assertEquals($size, (string) $xmlPart[0]);
```

Comment (jasson99) [reply tests/acceptance/features/bootstrap/WebDav.php]: fixed

File: tests/acceptance/features/bootstrap/WebDav.php (hunk 2 above)
Comment (jasson99) [reply tests/acceptance/features/bootstrap/WebDav.php]: fixed

File: tests/acceptance/features/bootstrap/WebDav.php (hunk 1 above)
Comment (jasson99) [reply tests/acceptance/features/bootstrap/WebDav.php]: fixed

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 4]
```diff
@@ -1099,6 +1099,58 @@ public function downloadFileAsUserUsingPassword(
...
+	/**
+	 * @When user :user gets size of file :resource shared by public link using the WebDAV API
+	 *
+	 * @param $user
+	 * @param $resource
+	 *
+	 * @return void
+	 * @throws Exception
+	 */
+	public function userGetsSizeOfFileSharedByPublicLinkUsingTheWebdavApi($user, $resource) {
```
[the remark this answers is not in the document]
Comment (jasson99) [reply tests/acceptance/features/bootstrap/WebDav.php]: fixed

File: tests/acceptance/features/bootstrap/WebDav.php [hunk 5]
```diff
@@ -1099,6 +1099,58 @@ public function downloadFileAsUserUsingPassword(
 		);
 	}
 
+	/**
+	 * @When user :user gets size of file :resource shared by public link using the WebDAV API
```
[the remark this answers is not in the document]
Comment (jasson99) [reply tests/acceptance/features/bootstrap/WebDav.php]: fixed

**Review by phil-davis — APPROVED** [review]

**Comment by phil-davis** [comment]
The acceptance tests here cover just the fixed `getcontentlength` in a `PROPFIND`
I raised issue #36821 to add tests to better cover data items returned by `PROPFIND` requests more generally.

**Comment by kiranparajuli589** [comment]
> I raised issue #36821 to add tests to better cover data items returned by `PROPFIND` requests more generally.

@phil-davis I donot understand your gist properly, what sorts of `data-items` you intend to? Please elaborate here so that I could add some scenarios...

**Comment by phil-davis** [comment]
Some stuff I can see:
`WebDavPropertiesContext.php`
```
	 * @When user :username gets the following properties of file/folder :path using the WebDav PropFind API
public function userGetsFollowingPropsWithNamespaceOfFileUsingWebDavAPI(...
```
That seems to be unused. It calls `WebDavHelper::propfindWithMultipleProps`
Maybe that is a useful step and method?

There is also:
```
	 * @When /^user "([^"]*)" gets the following properties of (?:file|folder|entry) "([^"]*)" using the WebDAV API$/
public function userGetsPropertiesOfFolder(...
```
What is the difference of these 2 methods? Why do we have 2 methods?

**Comment by phil-davis** [comment]
We have tests that check properties like:
`apiComments/comments.feature` checks properties like `oc:comments-href` `oc:comments-count` `oc:comments-unread`
`apiShareOperations/getWebDAVSharePermissions.feature` checks `ocs:share-permissions`
Other feature files check `d:lockdiscovery` attributes, `d:resourcetype` attributes, `oc:share-types`, `oc:privatelink`, `d:quota-available-bytes`

There are probably quite a few "interesting" properties returned.

e.g. https://doc.owncloud.com/server/developer_manual/webdav_api/comments.html

There are lots of properties returned about a comment. We can make some API test scenarios that verify that the properties do come...

And for the other APIs that return lists of properties, including the "general" webDAV API.

**Comment by kiranparajuli589** [comment]
> Some stuff I can see:
> `WebDavPropertiesContext.php`
> 
> ```
> 	 * @When user :username gets the following properties of file/folder :path using the WebDav PropFind API
> public function userGetsFollowingPropsWithNamespaceOfFileUsingWebDavAPI(...
> ```
> 
> That seems to be unused. It calls `WebDavHelper::propfindWithMultipleProps`
> Maybe that is a useful step and method?
> 
> There is also:
> 
> ```
> 	 * @When /^user "([^"]*)" gets the following properties of (?:file|folder|entry) "([^"]*)" using the WebDAV API$/
> public function userGetsPropertiesOfFolder(...
> ```
> 
> What is the difference of these 2 methods? Why do we have 2 methods?

@phil-davis Only difference between these two methods is they use different `depth` level for `PROPFIND` requests.

**Pull request merged**